README.md
# Unofficial wildfire data monitor
Gathers data from public sources, and posts it to Twitter.

[![Twitter Follow](https://img.shields.io/twitter/follow/FireUnofficial.svg?style=social)](https://twitter.com/FireUnofficial) [![Twitter Follow](https://img.shields.io/twitter/follow/CaliFireBot.svg?style=social)](https://twitter.com/CaliFireBot) [![Twitter Follow](https://img.shields.io/twitter/follow/SWFireBot.svg?style=social)](https://twitter.com/SWFireBot) [![Twitter Follow](https://img.shields.io/twitter/follow/PacNWFireBot.svg?style=social)](https://twitter.com/PacNWFireBot) [![Twitter Follow](https://img.shields.io/twitter/follow/NRockFireBot.svg?style=social)](https://twitter.com/NRockFireBot) [![Twitter Follow](https://img.shields.io/twitter/follow/UtahNevFireBot.svg?style=social)](https://twitter.com/UtahNevFireBot) [![Twitter Follow](https://img.shields.io/twitter/follow/RockyMtnFireBot.svg?style=social)](https://twitter.com/RockyMtnFireBot) [![Twitter Follow](https://img.shields.io/twitter/follow/EasternFireBot.svg?style=social)](https://twitter.com/EasternFireBot)  
[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](./LICENSE) [![Build Status](https://travis-ci.com/publicdocs-platform/fire-monitor-bot.svg?branch=master)](https://travis-ci.com/publicdocs-platform/fire-monitor-bot) [![GitHub last commit](https://img.shields.io/github/last-commit/publicdocs-platform/fire-monitor-bot.svg)](https://github.com/publicdocs-platform/fire-monitor-bot/commits) [![GitHub tag (latest SemVer)](https://img.shields.io/github/tag/publicdocs-platform/fire-monitor-bot.svg)](https://github.com/publicdocs-platform/fire-monitor-bot/releases) [![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)  
[![codecov](https://codecov.io/gh/publicdocs-platform/fire-monitor-bot/branch/master/graph/badge.svg)](https://codecov.io/gh/publicdocs-platform/fire-monitor-bot) [![Dependabot Status](https://api.dependabot.com/badges/status?host=github&repo=publicdocs-platform/fire-monitor-bot)](https://dependabot.com) [![Codacy Badge](https://api.codacy.com/project/badge/Grade/f2d4cb425efd4ef58a5926f7b973ea58)](https://app.codacy.com/app/publicdocs-platform/fire-monitor-bot?utm_source=github.com&utm_medium=referral&utm_content=publicdocs-platform/fire-monitor-bot&utm_campaign=Badge_Grade_Dashboard) [![DeepScan grade](https://deepscan.io/api/teams/4040/projects/5843/branches/46602/badge/grade.svg)](https://deepscan.io/dashboard#view=project&tid=4040&pid=5843&bid=46602) [![This project is using Percy.io for visual regression testing.](https://percy.io/static/images/percy-badge.svg)](https://percy.io/Public-Docs-Project/fire-monitor-bot)


The original runs at https://twitter.com/FireUnofficial

The generated database is archived and updated at: https://github.com/publicdocs/wildfire-data-archive

The fire information displayed by this app is UNOFFICIAL, FOR INFORMATION ONLY, 
NOT SUITABLE FOR SAFETY/EMERGENCY PURPOSES, 
and MAY BE INCORRECT OR OUT-OF-DATE. USE AT YOUR OWN RISK.


## Features

* Sources fire summary from the NWCG NFSA, USGS GeoMAC, and CAL FIRE
* Sources perimeter info from USGS GeoMAC
* Sources satellite fire detection from MODIS and VIIRS I (NASA/NOAA via Forest Service Active Fire Mapping)
* Creates a multi-layer base map from USGS, US Census Bureau and other federal govt data sets
* Determines location of fire relative to named populated places from the Census Bureau
* Monitors differences in fire information over time, and tweets out the changes (summary, perimeter, and satellite fire readings)


## License

Read the [LICENSE](LICENSE) for details.  
The entire [NOTICE](NOTICE) file serves as the NOTICE that must be included under
Section 4d of the License.

````
The fire-monitor-bot software is available under the license below.
Third-party dependencies not distributed with fire-monitor-bot
and data used during runtime may be under other licenses.

The fire information displayed by this app is
* UNOFFICIAL,
* FOR INFORMATION ONLY, 
* NOT SUITABLE FOR SAFETY/EMERGENCY/EVACUATION PURPOSES, 
* and MAY BE INCORRECT OR OUT-OF-DATE.

USE AT YOUR OWN RISK.  DANGER MAY BE CLOSER THAN SHOWN HERE.

Note that while we may distribute the images and posts generated and
published by our original bot using the disclaimers indicated in the
relevant template files, you must decide for yourself whether
you wish to use the same disclaimers and/or whether those disclaimers meet
your obligations under both our and third-party requirements.
We cannot give any legal advice on that.

No license to use any trademarks is provided.  You are not permitted
to use any of the names of our Twitter accounts in any content you
generate.

=====

fire-monitor-bot (https://github.com/publicdocs-platform/fire-monitor-bot)

Copyright (c) 2018-2019 Advay Mengle <source@madvay.com>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this software except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
````


CHANGELOG.md
## [3.7.61](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.60...v3.7.61) (2020-07-18)


### Bug Fixes

* **deps:** bump lodash from 4.17.15 to 4.17.19 ([#604](https://github.com/publicdocs-platform/fire-monitor-bot/issues/604)) ([e8ce242](https://github.com/publicdocs-platform/fire-monitor-bot/commit/e8ce242d082091170de647032c12230a78f9b554))

## [3.7.60](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.59...v3.7.60) (2020-07-18)


### Bug Fixes

* **deps:** bump puppeteer from 2.1.1 to 5.2.0 ([#616](https://github.com/publicdocs-platform/fire-monitor-bot/issues/616)) ([9495647](https://github.com/publicdocs-platform/fire-monitor-bot/commit/9495647e8f4f1744e3945e62aa387a429c632fe3))

## [3.7.59](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.58...v3.7.59) (2020-07-08)


### Bug Fixes

* **deps:** bump js-yaml from 3.13.1 to 3.14.0 ([5b178e5](https://github.com/publicdocs-platform/fire-monitor-bot/commit/5b178e548ad60aaef64ff7827becd6f32fe3fcb6))

## [3.7.58](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.57...v3.7.58) (2020-07-07)


### Bug Fixes

* **deps:** [security] bump npm from 6.13.4 to 6.14.6 ([#600](https://github.com/publicdocs-platform/fire-monitor-bot/issues/600)) ([2996574](https://github.com/publicdocs-platform/fire-monitor-bot/commit/29965741e1c52f36df45225e29372ad9d7fcefa1))

## [3.7.57](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.56...v3.7.57) (2020-05-22)


### Bug Fixes

* **deps:** bump snyk from 1.323.2 to 1.324.0 ([e83d13f](https://github.com/publicdocs-platform/fire-monitor-bot/commit/e83d13f9463f23ea40705bbf17fddb6890736f06))

## [3.7.56](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.55...v3.7.56) (2020-05-21)


### Bug Fixes

* **deps:** bump snyk from 1.323.1 to 1.323.2 ([432ba13](https://github.com/publicdocs-platform/fire-monitor-bot/commit/432ba13ffef18952070dfd3169f5ca7921001b24))

## [3.7.55](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.54...v3.7.55) (2020-05-20)


### Bug Fixes

* **deps:** bump snyk from 1.321.0 to 1.323.1 ([bba1ecd](https://github.com/publicdocs-platform/fire-monitor-bot/commit/bba1ecdadede9c6e26f8c01558ee0a4295a556bf))

## [3.7.54](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.53...v3.7.54) (2020-05-16)


### Bug Fixes

* **deps:** bump luxon from 1.22.2 to 1.24.1 ([1b8e155](https://github.com/publicdocs-platform/fire-monitor-bot/commit/1b8e155a467cf4901b262f927f0d862843661d09))

## [3.7.53](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.52...v3.7.53) (2020-05-16)


### Bug Fixes

* **deps:** [security] bump handlebars from 4.5.3 to 4.7.6 ([dcdfe9a](https://github.com/publicdocs-platform/fire-monitor-bot/commit/dcdfe9a33ac30b47c18bfe1d00250ca0d103fa00))

## [3.7.52](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.51...v3.7.52) (2020-05-14)


### Bug Fixes

* **deps:** bump proj4 from 2.6.0 to 2.6.1 ([1baed5f](https://github.com/publicdocs-platform/fire-monitor-bot/commit/1baed5f89ef30a76682b828f9407faae3bba44a6))

## [3.7.51](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.50...v3.7.51) (2020-05-14)


### Bug Fixes

* **deps:** bump snyk from 1.305.0 to 1.321.0 ([97e4195](https://github.com/publicdocs-platform/fire-monitor-bot/commit/97e4195f497bf74da3199c878f14122bdb553a81))

## [3.7.50](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.49...v3.7.50) (2020-03-25)


### Bug Fixes

* **deps:** bump luxon from 1.22.0 to 1.22.2 ([4693b87](https://github.com/publicdocs-platform/fire-monitor-bot/commit/4693b87f5cfcf289a8817990a00fa6ea4cdf8f98))

## [3.7.49](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.48...v3.7.49) (2020-03-25)


### Bug Fixes

* **deps:** bump snyk from 1.304.0 to 1.305.0 ([1358980](https://github.com/publicdocs-platform/fire-monitor-bot/commit/13589801c50ad37602f1ae3c8a566aac3aca319a))

## [3.7.48](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.47...v3.7.48) (2020-03-24)


### Bug Fixes

* **deps:** bump snyk from 1.303.2 to 1.304.0 ([9982465](https://github.com/publicdocs-platform/fire-monitor-bot/commit/99824651df550e58af13d75bd03fad7e5b347d73))

## [3.7.47](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.46...v3.7.47) (2020-03-24)


### Bug Fixes

* **deps:** bump snyk from 1.303.1 to 1.303.2 ([72c9ca7](https://github.com/publicdocs-platform/fire-monitor-bot/commit/72c9ca7df8ae5a155191987389b4232e1e4f38b7))

## [3.7.46](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.45...v3.7.46) (2020-03-23)


### Bug Fixes

* **deps:** bump snyk from 1.303.0 to 1.303.1 ([41f88e0](https://github.com/publicdocs-platform/fire-monitor-bot/commit/41f88e0635733b222fc12ae5c381fa4088b69ec8))

## [3.7.45](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.44...v3.7.45) (2020-03-22)


### Bug Fixes

* **deps:** bump csv from 5.3.1 to 5.3.2 ([fadb4cc](https://github.com/publicdocs-platform/fire-monitor-bot/commit/fadb4ccf8019f18ae5dcd90c039cb6a555d22266))

## [3.7.44](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.43...v3.7.44) (2020-03-22)


### Bug Fixes

* **deps:** bump snyk from 1.301.0 to 1.303.0 ([4fc27cb](https://github.com/publicdocs-platform/fire-monitor-bot/commit/4fc27cb3915e66bbd134ae84bbb41e81a6e39448))

## [3.7.43](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.42...v3.7.43) (2020-03-20)


### Bug Fixes

* **deps:** bump snyk from 1.297.2 to 1.301.0 ([5675c67](https://github.com/publicdocs-platform/fire-monitor-bot/commit/5675c67dd355341ca7832fb68308bfc585f99a74))

## [3.7.42](https://github.com/publicdocs-platform/fire-monitor-bot/compare/v3.7.41...v3.7.42) (2020-03-18)


### Bug Fixes


perimMap.js
/*
Copyright (c) 2018 Advay Mengle <source@madvay.com>.
See the LICENSE and NOTICE files for details.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

function showMap(centerX, centerY, zoom, detail, cities0) {
  let cities = cities0 || { closest: [], biggest: [] };
  let showAll = 'show:';

  for (let s = 0; s < 100; s++) {
    showAll = showAll + s + ',';
  }

  let Library = {
    USGS: {
      NatlMap: {
        Blank: {
          url: 'https://basemap.nationalmap.gov/arcgis/rest/services/USGSTNMBlank/MapServer',
          params: { layers: '', FORMAT: 'PNG' }
        },
        Imagery: {
          url: 'https://basemap.nationalmap.gov/arcgis/rest/services/USGSImageryOnly/MapServer',
          params: { layers: '', FORMAT: 'PNG' }
        },
        ImageryTiled: {
          tiled: true,
          url: 'https://basemap.nationalmap.gov/arcgis/rest/services/USGSImageryOnly/MapServer/tile/{z}/{y}/{x}',
          params: {}
        },
        ImageryTopo: {
          url: 'https://basemap.nationalmap.gov/arcgis/rest/services/USGSImageryTopo/MapServer',
          params: { layers: '', FORMAT: 'PNG' }
        },
        Polygons: {
          url: 'https://carto.nationalmap.gov/arcgis/rest/services/selectable_polygons/MapServer',
          params: { layers: '', FORMAT: 'PNG' }
        },
        Roads: {
          url: 'https://carto.nationalmap.gov/arcgis/rest/services/transportation/MapServer/',
          params: { layers: 'show:13,18,21,22,23,25,26,27,28,29,30,31,32,33,34,35,36', FORMAT: 'PNG' }
        },
        RoadsMediumScale: {
          url: 'https://carto.nationalmap.gov/arcgis/rest/services/transportation/MapServer/',
          params: { layers: 'show:13,18,25,26,27,28,29,30,31,32,33,34,35,36', FORMAT: 'PNG' }
        },
        RoadsLowScale: {
          url: 'https://carto.nationalmap.gov/arcgis/rest/services/transportation/MapServer/',
          params: { layers: 'show:13,25,26,36', FORMAT: 'PNG' }
        },
        TransportNotInCensus: {
          url: 'https://carto.nationalmap.gov/arcgis/rest/services/transportation/MapServer/',
          params: { layers: 'show:13,18,19,25,26,29,30,31,32,33,34,35,36', FORMAT: 'PNG' }
        },
        GovUnits: {
          url: 'https://carto.nationalmap.gov/arcgis/rest/services/govunits/MapServer/',
          params: { layers: 'exclude:17', FORMAT: 'PNG' }
        },
        GovUnitsSelectedLabels: {
          url: 'https://carto.nationalmap.gov/arcgis/rest/services/govunits/MapServer/',
          params: { layers: 'show:4,5,6,7,16', FORMAT: 'PNG' }
        },
        Names: {
          url: 'https://carto.nationalmap.gov/arcgis/rest/services/geonames/MapServer/',
          params: { layers: ''/*'show:0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22'*/, FORMAT: 'PNG' }
        },
      },
      ProtectedAreas: {
        SimpleDesignations: {
          url: 'https://gis1.usgs.gov/arcgis/rest/services/PADUS1_4/SimpleDesignationType/MapServer',
          params: { layers: '', FORMAT: 'PNG' },
          opacity: 0.3,
        }
      }
    },
    GEOMAC: {
      Fires: {
        url: 'https://wildfire.cr.usgs.gov/ArcGIS/rest/services/geomac_dyn/MapServer',
        params: {
          FORMAT: 'PNG',
          layers: 'show:0,1,2',
        },
        opacity: 0.5,
      },
    },
    Census: {
      Tiger: {
        States: {
          url: 'https://tigerweb.geo.census.gov/arcgis/rest/services/TIGERweb/State_County/MapServer',
          params: { layers: 'show:0,2,4,6,8,10,12,14,15,16', FORMAT: 'PNG' }
        },
        Roads: {
          url: 'https://tigerweb.geo.census.gov/arcgis/rest/services/TIGERweb/Transportation/MapServer',
          params: { layers: '', FORMAT: 'PNG' }
        },
        USLandmass: {
          url: 'https://tigerweb.geo.census.gov/arcgis/rest/services/TIGERweb/USLandmass/MapServer',
          params: { layers: '', FORMAT: 'JPEG', TRANSPARENT: 'false' /* Need the blue oceans */ }
        },
        Hydro: {
          url: 'https://tigerweb.geo.census.gov/arcgis/rest/services/TIGERweb/Hydro/MapServer',
          params: { layers: '', FORMAT: 'PNG' }
        },
        HydroPaths: {
          url: 'https://tigerweb.geo.census.gov/arcgis/rest/services/TIGERweb/Hydro/MapServer',
          params: { layers: 'show:0', FORMAT: 'PNG' }
        },
        HydroBodies: {
          url: 'https://tigerweb.geo.census.gov/arcgis/rest/services/TIGERweb/Hydro/MapServer',
          params: { layers: 'show:1,2', FORMAT: 'PNG' }
        },
      }
    },
  }

  function Alpha(config, opacity) {
    let ret = Object.assign({}, config);
    ret.opacity = opacity;
    return ret;
  }


  const textStyle = function (feature) {
    return new ol.style.Text({
      textAlign: 'left',
      textBaseline: 'bottom',
      font: '14px Roboto',
      text: feature.get('name'),
      fill: new ol.style.Fill({ color: '#000000' }),
      stroke: new ol.style.Stroke({ color: '#ffffff', width: 3 }),
      offsetX: 0,
      offsetY: 0,
    });
  };



  function perimVectorLayer() {
    const baseUrl = 'https://wildfire.cr.usgs.gov/arcgis/rest/services/geomac_dyn/MapServer/2'
    function styles(feat) {
      let sclr = 'rgba(255,0,0,1)';
      let fclr = 'rgba(255,0,0,0.03)';
      let sclr2 = 'rgba(255,160,0,1)';
      return [new ol.style.Style({
        fill: new ol.style.Fill({
          color: fclr
        }),

        stroke: new ol.style.Stroke({ color: sclr2, width: 3 }),

      }), new ol.style.Style({
        stroke: new ol.style.Stroke({ color: sclr, width: 2 }),
      })];
    }
    let source = tiledVectorLayer(baseUrl, 1024);
    return new ol.layer.Vector({
      source: source,
      style: styles,
      declutter: false,
    });
  }


  function perimFillVectorLayer() {
    const baseUrl = 'https://wildfire.cr.usgs.gov/arcgis/rest/services/geomac_dyn/MapServer/2'
    function style(feat) {
      let sclr = 'rgba(255,0,0,0.7)';
      let fclr = 'rgba(255,255,0,0.5)';
      return new ol.style.Style({
        fill: new ol.style.Fill({
          color: fclr
        }),
        stroke: new ol.style.Stroke({ color: sclr, width: 2 }),

      });
    }
    let source = tiledVectorLayer(baseUrl, 1024);
    return new ol.layer.Vector({
      source: source,
      style: style,
      declutter: false,
    });
  }

  function satelliteVectorLayer(l) {
    const baseUrl = 'https://wildfire.cr.usgs.gov/arcgis/rest/services/geomac_dyn/MapServer/' + l;
    function style(feat) {
      let sclr = 'rgba(255,255,0,0.01)';
      let fclr = 'rgba(255,255,0,0)';
      const time = feat.get('load_stat');
      if (time == 'Last 24-48 hrs') {
        sclr = 'rgba(255,255,0,0.1)'
        fclr = 'rgba(255,255,0,0)'
      } else if (time == 'Last 12-24 hrs') {
        sclr = 'rgba(255,165,0,0.2)'
        fclr = 'rgba(255,165,0,0.1)'
      } else if (time == 'Active Burning') {
        sclr = 'rgba(255,0,0,0.4)'
        fclr = 'rgba(255,0,0,0.4)'
      } else {
        return null;
      }
      return new ol.style.Style({
        zIndex: -(feat.get('julian') * 1000000 + feat.get('gmt')),
        fill: zoom > 12.5 ? null : new ol.style.Fill({
          color: fclr
        }),

        stroke: new ol.style.Stroke({ color: sclr, width: zoom>=12.5 ? 3 : 1, lineDash:zoom>=12.5?[3,3]:null }),

      });
    }
    
    let source = tiledVectorLayer(baseUrl, 1024);
    return new ol.layer.Vector({
      source: source,
      style: style,
      declutter: false,
    });


  }

  function fireVectorLayer(l) {

    const textStyle = function (feature) {
      return new ol.style.Text({
        textAlign: 'center',
        textBaseline: 'bottom',
        font: '15px Roboto',
        text: feature.get('incidentname'),
        fill: new ol.style.Fill({ color: '#000000' }),
        stroke: new ol.style.Stroke({ color: '#ff0000', width: 3 }),
        offsetX: 0,
        offsetY: 0,
      });
    };
    
    function style(feat) {
      return new ol.style.Style({
        image:
          new ol.style.Circle({
            radius: 3,
            fill: new ol.style.Fill({
              color: 'rgba(255, 0, 0, 0.5)'
            }),
          })
      });
    }

    const baseUrl = 'https://wildfire.cr.usgs.gov/arcgis/rest/services/geomac_dyn/MapServer/' + l;
    
    let source = tiledVectorLayer(baseUrl, 1024);
    return new ol.layer.Vector({
      source: source,
      style: style,
      declutter: false,
    });


  }

  function citiesVectorLayer() {
    function cityFeature(city) {

      return {
        'type': 'Feature',
        'properties': {
          'name': city.name,
          'population': city.population,
        },
        'geometry': {
          'type': 'Point',
          'coordinates': [city.lon, city.lat],
        }
      };
    }

    let citiesBest = cities.closest;
    if (!detail) {
      citiesBest = cities.closest.slice(0, 3).concat(cities.biggest.slice(0, 5));
    }

    let allFeats = {
      'type': 'FeatureCollection',
      'crs': {
        'type': 'name',
        'properties': {
          'name': 'EPSG:4326'
        }
      },
      'features': citiesBest.map(cityFeature)
    };
    function style(feat) {
      return new ol.style.Style({
        zIndex: -feat.get('population'),
        text: textStyle(feat),
        image:
          new ol.style.Circle({
            radius: 3,
            fill: new ol.style.Fill({
              color: 'rgba(0, 0, 0, 1)'
            }),
          })
      });
    }
    return new ol.layer.Vector({
      source: new ol.source.Vector({
        features: (new ol.format.GeoJSON()).readFeatures(allFeats),
      }),
      style: style,
      declutter: true,
    });
  }

  let ZoomedRoads = zoom < 11.5 ? Library.USGS.NatlMap.RoadsLowScale : (zoom < 12.5 ? Library.USGS.NatlMap.RoadsMediumScale : Library.USGS.NatlMap.Roads);

  let perimLayers = [
    // Library.Census.Tiger.USLandmass,
    Library.USGS.NatlMap.Blank,
    Alpha(Library.USGS.NatlMap.ImageryTiled, 0.1),
    Alpha(Library.USGS.ProtectedAreas.SimpleDesignations, 0.1),
    //Library.USGS.NatlMap.Polygons,
    Library.Census.Tiger.States,
    Library.Census.Tiger.HydroBodies,
    //Alpha(Library.Census.Tiger.HydroPaths, 0.2),
    ZoomedRoads,
    'Perim',
    'VIIRS',
    'MODIS',
    Library.USGS.NatlMap.GovUnits,
    Library.Census.Tiger.Roads,
    Library.USGS.NatlMap.GovUnitsSelectedLabels,
    Library.USGS.NatlMap.Names,
  ];
  let overviewLayers = [
    Library.Census.Tiger.USLandmass,
    Library.USGS.NatlMap.Blank,
    Alpha(Library.USGS.NatlMap.ImageryTiled, 0.1),
    Alpha(Library.USGS.ProtectedAreas.SimpleDesignations, 0.1),
    Library.Census.Tiger.States,
    Library.Census.Tiger.Hydro,
    ZoomedRoads,
    Library.Census.Tiger.Roads,
    Library.USGS.NatlMap.GovUnitsSelectedLabels,
    'PerimFill',
    'Fires',
    'Complexes',
    Library.USGS.NatlMap.Names,
    'Cities',
  ];

  let configs = detail ? perimLayers : overviewLayers;
  
  function configToLayer(config) {
    let source = null;
    let ltype = null;
    if (config === 'VIIRS') {
      return satelliteVectorLayer(5);
    } else if (config === 'MODIS') {
      return satelliteVectorLayer(4);
    } else if (config === 'Perim') {
      return perimVectorLayer();
    } else if (config === 'PerimFill') {
      return perimFillVectorLayer();
    } else if (config === 'Cities') {
      return citiesVectorLayer();
    } else if (config === 'Fires') {
      return fireVectorLayer(0);
    } else if (config === 'Complexes') {
      return fireVectorLayer(1);
    }
    if (config.tiled) {
      ltype = ol.layer.Tile;
      source = new ol.source.XYZ({
        hidpi: detail || true,
        ratio: 1,
        params: config.params || {},
        url: config.url,
      });
    } else {
      ltype = ol.layer.Image;
      source = new ol.source.ImageArcGISRest({
        hidpi: detail || true,
        ratio: 1,
        params: config.params || {},
        url: config.url,
      });
    }

    return new ltype({
      source: source,
      opacity: config.opacity || 1.0,
    });
  }

  let layers = configs.map(configToLayer);
  let controls = [];

  controls.push(new ol.control.ScaleLine({ units: 'us', minWidth: 70 }));


  let map = new ol.Map({
    controls: controls,
    pixelRatio: detail ? 3 : 3,
    layers: layers,
    target: 'map',
    view: new ol.View({
      projection: 'EPSG:4326',
      center: [centerX, centerY],
      zoom: zoom
    })
  });
}

